astropy/modeling/fitting.py

```diff
@@ -618,10 +618,12 @@ def __call__(self, model, x, y, z=None, weights=None, **kwargs):
 
         Returns
         -------
-        filtered_data : numpy.ma.core.MaskedArray
```

Comment:
This looks like a significant API change. Won't this break existing code on user side?

Reply:
Probably, but that was @nden's preference (#7350) and avoids returning an unnecessary (and currently sort-of-corrupted) copy of the input data. Just to be clear, I don't have a pressing need for this change; I'm just helping wrap up the changes from the discussion in #7199.

Reply:
I suppose the painful thing here would be how to support both AstroPy >= 3.1 and earlier versions in application code. I'm thinking that something like this would be necessary in the most general case, which is a bit ugly:
```
fitted_model, mask = fitting_with_outlier_removal_instance(args)

if isinstance(fitted_model, np.ndarray):
    fitted_model, mask = mask, fitted_model
if isinstance(mask, np.ma.masked_array):
    mask = mask.mask
elif mask.dtype != np.bool:  # bad return type for niter=0 in old versions:
    mask = np.zeros_like(mask, dtype=np.bool)
```
But I think Nadia's reasoning was that it's better to make such changes now than later.

Reply:
I feel uncomfortable with sudden API change without a deprecation period, but the decision is beyond me. I wonder if astropy/astropy-APEs#20 applies here. :thinking:

Reply:
@pllim The modeling subpackage has a Warning (note) that it is a work in progress and the API may change. The goal for the next release is to stabilize the API and remove the Warning. I expect to have other API changes. I am taking that `Note` as equivalent to deprecation warning. Is there a better way to handle this situation?

Reply:
As discussed offline at pyastro, I think we should think about this carefully with stakeholders.